File: swish.py
import math
from torch import nn
from torch.autograd import Function
import torch

import swish_cpp

torch.manual_seed(42)


class SWISHFunction(Function):
    @staticmethod
    def forward(ctx, x, beta):
        output = swish_cpp.forward(x, beta)[0]
        variables = [x, beta]
        ctx.save_for_backward(*variables)

        return output

    @staticmethod
    def backward(ctx, grad_y):
        grad_x, grad_beta = swish_cpp.backward(grad_y, *ctx.saved_variables)
        return grad_x, grad_beta
        # grad_x = swish_cpp.backward(grad_y, *ctx.saved_variables)[0]
        # return grad_x, None


class SWISH(nn.Module):
    def __init__(self):
        super(SWISH, self).__init__()
        # self.input_feature = input_feature
        # self.output_feature = output_feature
        self.beta = nn.Parameter(torch.Tensor(1))
        self.beta.data.uniform_(0, 1.5)

    def forward(self, x):
        return SWISHFunction.apply(x, self.beta)

File: swish.cpp
#include <torch/extension.h>

#include <iostream>

#include <vector>

torch::Tensor swish(torch::Tensor z, torch::Tensor beta) {
  auto s = torch::sigmoid(beta*z);
  return z * s;
}

std::vector<at::Tensor> swish_forward(
    torch::Tensor x,
    torch::Tensor beta) {
  auto new_x = x*torch::sigmoid(beta*x);

  return {new_x};
}

std::vector<torch::Tensor> swish_backward(
    torch::Tensor grad_y,
    torch::Tensor x,
    torch::Tensor beta) {
  auto grad_x = beta*swish(x, beta) + torch::sigmoid(beta*x)*(1-beta*swish(x, beta));
  auto grad_beta =x*x*torch::sigmoid(beta*x)*(1-torch::sigmoid(beta*x));

  return {grad_x, grad_beta};
}
